### PR TITLE
Do not serialize extern aliases to synthesized method

### LINKED ISSUE #63927: PdbWriter.SerializeDebugInfo() drops extern aliases if the first method is a synthesized method
See [PdbWriter.SerializeDebugInfo()](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs#L96).

### PR DESCRIPTION
**Pull request by cston**
`PdbWriter.SerializeDebugInfo()` drops extern aliases from the native PDB if the first method has no other debug info.

Fixes #63927.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
I wasn't sure what "extern namespace" means here. Also, I'm not sure I understand the impact of the change. It looks like this might cause us to produce debug info more often than we otherwise would.

**Review by tmat — APPROVED** [review]

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review (this change is blocking https://github.com/dotnet/roslyn/pull/63916), thanks.

File: src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs [hunk 1]
```diff
@@ -123,6 +103,27 @@ public bool ShouldForwardNamespaceScopes(EmitContext context, IMethodBody method
 
             byte[] result = encoder.ToArray() ?? Array.Empty<byte>();
             pooledBuilder.Free();
+
+            emitExternNamespaces = false;
+
+            // CONSIDER: this may not be the same "first" method as in Dev10, but
+            // it shouldn't matter since all methods will still forward to a method
+            // containing the appropriate information.
+            if (_methodBodyWithModuleInfo == null && result.Length > 0)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs:112]: >result.Length > 0 [](http://example.com/codeflow?start=53&length=17)

It looks like this condition makes a difference at the call site only when ```emitAllDebugInfo``` is false, and, it looks like in that case we are not interested in emitting information about extern aliases even if the blob was not empty. Here we are applying the filter unconditionally. This feels fragile to me, it is not obvious we are not going to break something. If our goal is to skip synthesized methods, perhaps we can filter those out explicitly here, or in the caller?

Comment (cston) [reply src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs:112]: The `Length` check is to ensure we're only notifying the caller to emit extern aliases if we've generated some other debug info here. The goal is to skip methods where no debug info is needed, not necessarily synthesized methods.

[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs:112]: Updated the condition and added a comment.

Comment (cston) [reply src/Compilers/Core/Portable/PEWriter/CustomDebugInfoWriter.cs:112]: Good catch, this is more fragile than I'd thought. Changed to check for `emitStateMachineInfo` directly and added an assert in the caller.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**